packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -20,64 +23,417 @@ const double _kHandleSize = 22.0;
   final VoidCallback handleSelectAll;
 
+  // When true, the toolbar fits above its anchor and will be positioned there.
+  final bool isAbove;
+
+  @override
+  _TextSelectionToolbarState createState() => _TextSelectionToolbarState();
+}
+
+class _TextSelectionToolbarState extends State<_TextSelectionToolbar> with TickerProviderStateMixin {
+  final GlobalKey _key = GlobalKey();
+
+  // Whether or not the overflow menu is open.  When it is closed, the menu
+  // items that don't overflow are shown. When it is open, only the overflowing
+  // menu items are shown.
+  bool _overflowOpen = false;
+  // The greatest width ever measured for this widget. This is used to position
+  // the overflow menu when it is open, since it is usually not as wide as the
+  // closed menu.
+  double _maxWidth;
+
+  FlatButton _getItem(VoidCallback onPressed, String label) {
+    if (onPressed == null) {
+      return null;
+    }
+    return FlatButton(
+      child: Text(label),
+      onPressed: () {
+        setState(() {
+          _overflowOpen = false;
+        });
+        onPressed();
+      },
+    );
+  }
+
   @override
   Widget build(BuildContext context) {
+    // Keep track of the largest child width.
+    if (_key.currentContext != null) {
+      final RenderBox renderBoxChild = _key.currentContext.findRenderObject() as RenderBox;
+      if (_maxWidth == null || renderBoxChild.size.width > _maxWidth) {
+        _maxWidth = renderBoxChild.size.width;
+      }
+    }
+
     final MaterialLocalizations localizations = MaterialLocalizations.of(context);
     final List<Widget> items = <Widget>[
-      if (handleCut != null) FlatButton(child: Text(localizations.cutButtonLabel), onPressed: handleCut),
-      if (handleCopy != null) FlatButton(child: Text(localizations.copyButtonLabel), onPressed: handleCopy),
-      if (handlePaste != null) FlatButton(child: Text(localizations.pasteButtonLabel), onPressed: handlePaste),
-      if (handleSelectAll != null) FlatButton(child: Text(localizations.selectAllButtonLabel), onPressed: handleSelectAll),
+      if (widget.handleCut != null)
+        _getItem(widget.handleCut, localizations.cutButtonLabel),
+      if (widget.handleCopy != null)
+        _getItem(widget.handleCopy, localizations.copyButtonLabel),
+      if (widget.handlePaste != null)
+        _getItem(widget.handlePaste, localizations.pasteButtonLabel),
+      if (widget.handleSelectAll != null)
+        _getItem(widget.handleSelectAll, localizations.selectAllButtonLabel),
     ];
 
     // If there is no option available, build an empty widget.
     if (items.isEmpty) {
       return Container(width: 0.0, height: 0.0);
     }
 
-    return Material(
-      elevation: 1.0,
-      child: Container(
-        height: _kToolbarHeight,
-        child: Row(mainAxisSize: MainAxisSize.min, children: items),
+    final Container child = Container(
+      key: _key,
+      height: _overflowOpen ? null : _kToolbarHeight,
+      child: Material(
```

Comment:
Note: Consider drawing this inside of _TextSelectionToolbarItems. Then you can do the calculation you're doing with `_maxWidth` inside of _TextSelectionToolbarItems.

Reply:
Not needed if the approach in https://github.com/flutter/flutter/pull/49391/commits/42f976714798555c8e95e8708a499b6a1fd229fe is ok.